src/dune_rules/context_env.ml

```diff
@@ -0,0 +1,225 @@
+        let fn = Path.relative dir prog in
+        let+ exists = Fs_memo.file_exists (Path.as_outside_build_dir_exn fn) in
+        if exists then Some fn else None)
+end
+
+let best_path =
+  let programs_for_which_we_prefer_opt_ext =
+    [ "ocamlc"; "ocamldep"; "ocamlmklib"; "ocamlobjinfo"; "ocamlopt" ]
+  in
+  fun ~dir program ->
+    let exe_path program =
+      let fn = Path.relative dir (program ^ Bin.exe) in
+      let+ exists = Fs_memo.file_exists (Path.as_outside_build_dir_exn fn) in
+      if exists then Some fn else None
+    in
+    if List.mem programs_for_which_we_prefer_opt_ext program ~equal:String.equal
+    then
+      let* path = exe_path (program ^ ".opt") in
+      match path with
+      | None -> exe_path program
+      | Some _ as path -> Memo.return path
+    else exe_path program
+
+module rec Rec : sig
+  val which : path:Path.t list -> string -> Path.t option Memo.t
+end = struct
+  open Rec
+
+  let which_impl (path, program) =
+    match path with
+    | [] -> Memo.return None
+    | dir :: path -> (
+      let* res = best_path ~dir program in
+      match res with
+      | None -> which ~path program
+      | Some prog -> Memo.return (Some prog))
+
+  let which =
+    let memo =
+      let module Input = struct
+        type t = Path.t list * string
+
+        let equal = Tuple.T2.equal (List.equal Path.equal) String.equal
+
+        let hash = Tuple.T2.hash (List.hash Path.hash) String.hash
+
+        let to_dyn = Dyn.opaque
+      end in
+      Memo.create "which"
+        ~input:(module Input)
+        ~cutoff:(Option.equal Path.equal) which_impl
+    in
+    fun ~path prog -> Memo.exec memo (path, prog)
+end
+
+include Rec
+
+let extend_paths t ~env =
+  let t =
+    let f (var, t) =
+      let parse ~loc:_ s = s in
+      let standard = Env_path.path env |> List.map ~f:Path.to_string in
+      (var, Ordered_set_lang.eval t ~parse ~standard ~eq:String.equal)
+    in
+    List.map ~f t
+  in
+  let vars =
+    let to_absolute_filename s =
+      Path.of_string s |> Path.to_absolute_filename
+    in
+    let sep = String.make 1 Bin.path_sep in
+    let env = Env.Map.of_list_exn t in
+    let f l = String.concat ~sep (List.map ~f:to_absolute_filename l) in
+    Env.Map.map ~f env
+  in
+  Env.extend ~vars env
+
+module Opam : sig
+  (** Environment for this opam switch *)
+  val env :
+    env:Env.t -> root:string option -> switch:string -> string Env.Map.t Memo.t
+end = struct
+  let opam =
+    Memo.Lazy.create ~name:"context-opam" (fun () ->
+        Bin.which ~path:(Env_path.path Env.initial) "opam" >>= function
+        | None -> Utils.program_not_found "opam" ~loc:None
+        | Some opam -> (
+          let+ version =
+            Memo.of_reproducible_fiber
+              (Process.run_capture_line ~display:Quiet Strict opam
+                 [ "--version"; "--color=never" ])
+          in
+          match Scanf.sscanf version "%d.%d.%d" (fun a b c -> (a, b, c)) with
```

Comment:
Will this break if opam returns `2.2.0-alpha1`?

Reply:
No, `Scanf.sscanf` will happily ignore everything in the input after it has finished scanning. That also means that `2.2.0~alpha1` and `2.2.0` will return the exact same `(2, 2, 0)` tuple.